### PR TITLE
Estimator RNN Model with Attention

### PR DESCRIPTION
**Pull request by jjtan**
Have not started model directory / structure refactor. This is just to get a plain estimator example in.

### CODE REVIEW

File: experiments/tf_trainer/tf_rnn_model.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        config=run_config)
+    return estimator
+
+  def _model_fn(self, features, labels, mode, params, config):
+    inputs = features[self._text_feature_name]
+    batch_size = tf.shape(inputs)[0]
+
+    rnn_layers = [
+        tf.nn.rnn_cell.GRUCell(num_units=size, activation=tf.nn.tanh)
+        for size in [128]
```
Comment (nthain) [inline experiments/tf_trainer/tf_rnn_model.py]: Let's make this a hyper-parameter the rnn_cell_sizes a hyperparameter.

File: experiments/tf_trainer/tf_rnn_model.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    # create a RNN cell composed sequentially of a number of RNNCells
+    multi_rnn_cell = tf.nn.rnn_cell.MultiRNNCell(rnn_layers)
+
+    outputs, states = tf.nn.dynamic_rnn(
+        multi_rnn_cell,
+        inputs,
+        sequence_length=tf.fill(dims=[batch_size], value=params.max_seq_length),
+        dtype=tf.float32)
+    outputs = tf.reshape(outputs, [batch_size, params.max_seq_length, 128])
```
Comment (nthain) [inline experiments/tf_trainer/tf_rnn_model.py]: I don't think this reshape is necessary as the output should be this shape already according to the tf.nn.dynamic_rnn docs: https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn

In any case, if you do reshape, please integrate with rnn_cell_sizes as a hyperparameter.

File: experiments/tf_trainer/tf_rnn_model.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                ],
+                axis=1),
+            axis=-1), -1)
+
+    weighted_output = tf.multiply(attention, outputs)
+    weighted_output = tf.reduce_sum(weighted_output, -2)
+
+    logits = tf.layers.dense(
+        inputs=weighted_output, units=128, activation=tf.nn.relu)
+    logits = tf.layers.dropout(logits, rate=0.3)
```
Comment (nthain) [inline experiments/tf_trainer/tf_rnn_model.py]: Again, let's make the dropout rate a hparam.

File: experiments/tf_trainer/tf_rnn_model.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                        units=1,
+                        activation=None) for output in unstacked_outputs
+                ],
+                axis=1),
+            axis=-1), -1)
+
+    weighted_output = tf.multiply(attention, outputs)
+    weighted_output = tf.reduce_sum(weighted_output, -2)
+
+    logits = tf.layers.dense(
```
Comment (nthain) [inline experiments/tf_trainer/tf_rnn_model.py]: Perhaps we can make this a hyperparameter as well (final_fully_connected_layer_sizes)

File: experiments/tf_trainer/tf_rnn_model.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      # Provide an estimator spec for `ModeKeys.TRAIN` modes.
+      return tf.estimator.EstimatorSpec(mode=mode, loss=loss, train_op=train_op)
+
+    if mode == tf.estimator.ModeKeys.EVAL:
+      probabilities = tf.sigmoid(logits)
+      guesses = tf.to_int32(probabilities > 0.5)
+
+      eval_metric_ops = {
+          'accuracy': tf.metrics.accuracy(labels, guesses),
+          'auroc': tf.metrics.auc(labels, probabilities)
```
Comment (nthain) [inline experiments/tf_trainer/tf_rnn_model.py]: Is auroc intentional? I've usually seen it as auc_roc or aucroc.

File: experiments/tf_trainer/tf_rnn_model.py (hunk 1 above)
Comment (jjtan) [reply experiments/tf_trainer/tf_rnn_model.py]: Let me defer all hyper parameter stuff. Going to go through and set up the hparams for ml engine in a later PR so will do all that then.

File: experiments/tf_trainer/tf_rnn_model.py (hunk 2 above)
Comment (jjtan) [reply experiments/tf_trainer/tf_rnn_model.py]: Yeah the reshape isn't doing anything but I had to include it to help TF know the shape for some reason. Let me add a comment.
